src/main/java/org/zalando/nakadi/validation/JsonSchemaLoader.java

```diff
@@ -0,0 +1,191 @@
+package org.zalando.nakadi.validation;
+
+import com.google.common.collect.Lists;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.zalando.nakadi.domain.CompatibilityMode;
+import org.zalando.nakadi.domain.EventType;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public class JsonSchemaLoader {
+    public static final String DATA_CHANGE_WRAP_FIELD = "data";
+
+    private static final String ADDITIONAL_PROPERTIES = "additionalProperties";
+    private static final String ADDITIONAL_ITEMS = "additionalItems";
+    private final List<String> keywordsWithNestedSchemas = Lists.newArrayList("definitions", "dependencies",
+            "properties");
+    private final List<String> keywordsInferObjectSchema = Lists.newArrayList("required", "minProperties",
+            "maxProperties");
+    private final List<String> keywordsInferArraySchema = Lists.newArrayList("minItems", "maxItems", "uniqueItems");
+    private final List<String> keywordsWithArrayOfSchemas = Lists.newArrayList("anyOf", "allOf", "oneOf");
+
+    public JSONObject effectiveSchema(final EventType eventType) throws JSONException {
+        final JSONObject schema = new JSONObject(eventType.getSchema().getSchema());
+
+        if (eventType.getCompatibilityMode().equals(CompatibilityMode.COMPATIBLE)) {
+            this.enforceStrictValidation(schema);
+        }
+
+        switch (eventType.getCategory()) {
+            case BUSINESS: return addMetadata(schema, eventType);
+            case DATA: return wrapSchemaInData(schema, eventType);
+            default: return schema;
+        }
+    }
+
+    private void enforceStrictValidation(final JSONObject schema) {
```

Comment:
This method forces strict validation for events. We describe in the concept document (linked in the description) the reasons for doing so. What strict means is that, only attributes that are explicitly defined in the json schema are going to be allowed. Anything else will cause events to be rejected.

Strict validation is only used by "compatible" event types.